### PR TITLE
When transfering repository and database transaction failed, rollback the renames

### LINKED ISSUE #14821: Repo Transfer: If the sql transaction fails we will be left repository in the wrong place
If the transaction fails we will be left repository in the wrong place

_Originally posted by @zeripath in https://github.com/go-gitea/gitea/pull/14792#discussion_r584185299_

### PR DESCRIPTION
**Pull request by lunny**
Fix #14821

### CODE REVIEW

File: models/repo_transfer.go [hunk 1]
```diff
@@ -305,10 +305,12 @@ func TransferOwnership(doer *User, newOwnerName string, repo *Repository) error
...
-	if isExist, err := util.IsExist(wikiPath); err != nil {
+	isExist, err := util.IsExist(wikiPath)
+	if err != nil {
 		log.Error("Unable to check if %s exists. Error: %v", wikiPath, err)
 		return err
-	} else if isExist {
+	}
+	if isExist {
 		if err := os.Rename(wikiPath, WikiPath(newOwner.Name, repo.Name)); err != nil {
 			return fmt.Errorf("rename repository wiki: %v", err)
```
Comment (zeripath) [inline models/repo_transfer.go:348]: Need to move the repository back here

**Comment by zeripath** [comment]
This probably also needs a panic handler too. 

My suspicion is that putting the return in a defer would be better:

e.g.
```
func TransferOwnership(doer *User, newOwnerName string, repo *Repository) (err error) {
   repoRenamed := false
   wikiRenamed := false
  
   defer func() {
       recErr := recover()

       if recErr != nil || err != nil {
          // handle moving stuff back
          // If there is an error moving stuff back this needs a clear log.Critical(...) so administrators know that they need to do something
      } ()
      ...
   }
```

Then you can just err = sess.Commit() and return err 

Something like that.

**Comment by lunny** [comment]
> This probably also needs a panic handler too.
> 
> My suspicion is that putting the return in a defer would be better:
> 
> e.g.
> 
> ```
> func TransferOwnership(doer *User, newOwnerName string, repo *Repository) (err error) {
>    repoRenamed := false
>    wikiRenamed := false
>   
>    defer func() {
>        recErr := recover()
> 
>        if recErr != nil || err != nil {
>           // handle moving stuff back
>           // If there is an error moving stuff back this needs a clear log.Critical(...) so administrators know that they need to do something
>       } ()
>       ...
>    }
> ```
> 
> Then you can just err = sess.Commit() and return err
> 
> Something like that.

It's running beyond on http router, and I think we have already recovery middleware there.

**Comment by zeripath** [comment]
yes but if there is a panic after the initial rename has occurred but then you're not going to rename back unless you use a recover deferral scheme as I suggest above.

File: models/repo_transfer.go [hunk 2]
```diff
@@ -305,10 +305,12 @@ func TransferOwnership(doer *User, newOwnerName string, repo *Repository) error
 	// Rename remote wiki repository to new path and delete local copy.
 	wikiPath := WikiPath(oldOwner.Name, repo.Name)
 
-	if isExist, err := util.IsExist(wikiPath); err != nil {
+	isExist, err := util.IsExist(wikiPath)
```
Comment (6543) [inline models/repo_transfer.go]: can we rename it to wikiExist if we widen scope to whole func?

**Comment by zeripath** [comment]
See https://github.com/lunny/gitea/pull/14

File: models/repo_transfer.go [hunk 3]
```diff
@@ -301,17 +335,21 @@ func TransferOwnership(doer *User, newOwnerName string, repo *Repository) error
...
+	if err != nil {
 		log.Error("Unable to check if %s exists. Error: %v", wikiPath, err)
 		return err
-	} else if isExist {
+	}
+	if isExist {
```
Comment (6543) [inline models/repo_transfer.go]:
```suggestion
	} else if isExist {
```

File: models/repo_transfer.go [hunk 4]
```diff
@@ -301,17 +335,21 @@ func TransferOwnership(doer *User, newOwnerName string, repo *Repository) error
...
 		return fmt.Errorf("rename repository directory: %v", err)
 	}
+	repoRenamed = true
 
 	// Rename remote wiki repository to new path and delete local copy.
 	wikiPath := WikiPath(oldOwner.Name, repo.Name)
 
-	if isExist, err := util.IsExist(wikiPath); err != nil {
+	isExist, err := util.IsExist(wikiPath)
+	if err != nil {
```
Comment (6543) [inline models/repo_transfer.go]:
```suggestion
	if isExist, err := util.IsExist(wikiPath); err != nil {
```

**Review by zeripath — APPROVED** [review]

**Review by 6543 — APPROVED** [review]

**Pull request merged**